### ISSUE TITLE
Update project to use Node.js v18.x

### ISSUE THREAD

**Issue by olof-nord** [mastodon/mastodon#20968]
### Pitch

Upgrade the project to the latest Node.js LTS release, v18.

### Motivation

This latest release is following the old LTS, v16 and contains the latest and greatest of security, features and is the way forward for the Node.js ecosystem.

**Comment by olof-nord** [comment]
Here is a summary of the current state of affairs.

The main (only?) issue with upgrading to Node.js v18 is that it is removing support for the now basically ancient MD4 hashing algorithm.

Currently, the project is using Webpack 4: this is no longer supported and instead superseded by v5.
Alongside this, we are also using a few different Webpack 4 loaders and plugins.

It is possible to re-configure webpack 4 to use a different hashing algorithm such as sha256, however this is not possible for the loaders and the plugins.

One example of a plugin which has hard-coded their "md4" usage is `compression-webpack-plugin` v6.x - see this issue for all details: https://github.com/webpack-contrib/compression-webpack-plugin/issues/329

One example of a loader which also hard-codes its hashing algorithm, assuming md4 is available, is `file-loader` v6.x.x.

**Comment by olof-nord** [comment]
The reason why Webpack 4 is used, is as it is the last version supported by Webpacker, the ruby gem.

Webpacker only supports webpack 4.x.x and will not be updated to newer versions.

My only conclusion here is that for this project to support Node.js v18, we need to replace Webpacker. Once we have replaced Webpacker, we can upgrade to Webpack v5, allowing us to upgrade our Webpack loaders and plugins.

Or, we move on to Node.js, but enable the the Node `--openssl-legacy-provider` flag, however I see that option as the worse of the two.

**Comment by olof-nord** [comment]
Here is a list of all loaders and plugin used by the project, alongside their usage of 'md4'.
The versions checked are the ones currently listed in master.

Here is a [link](https://github.com/mastodon/mastodon/blob/dde60e2634072753e1a83ee9cc3b8e6ff3f43202/package.json) with the latest commit as of me checking.

| Name         | uses md4     |
|--------------|-----------|
| compression-webpack-plugin | yes      |
| terser-webpack-plugin      | yes, but can be reconfigured  |
| webpack-bundle-analyzer      | no  |
| mini-css-extract-plugin      | no  |
| webpack-assets-manifest      | no, it uses sha256  |
| file-loader      | yes  |
| babel-loader      | yes, but can be reconfigured  |
| css-loader      | yes |
| postcss-loader      | no? |
| sass-loader      | no? |
| workbox-webpack-plugin      | no |

**Comment by olof-nord** [comment]
I put together a small demo project as to allow testing out the different libraries separated from this project here:

https://gitlab.com/olof-nord/webpack4-node18-demo

**Comment by renchap** [comment]
Replacing webpacker might be a good idea as it is no longer maintained.

You can either use your own Webpack integration, or Shakapacker, or decide do stop using Webpack and switch to [Vite Ruby](https://vite-ruby.netlify.app).

I recently migrated my app from Webpacker 6 beta (which was using Webpack 5) to Vite Ruby and it works really well, the compilation times are much better and even if I had a quite optimized setup with Webpack, `esbuild` is doing some magic and my JS bundles are 20-30% smaller now.

I think I can start working on migrating to Vite Ruby if there is an interest in a PoC, but I wont start working on it until it gains interest, as it will probably be a non-trivial amount of work.

**Comment by nicolo-ribaudo** [comment]
> Name | uses md4
> -- | --
> ... | ...
> babel-loader | yes, but can be reconfigured

`babel-loader` automatically switches to a different algorithm when using Node.js >= 17, so it doesn't even need to be reconfigured :)

**Comment by renchap** [comment]
> Just to try to make the vite plan a bit more concrete: that would mean completely replace webpack, loaders and plugins in the JavaScript frontend part and at the same time replace Webpacker with Vite Ruby?

Correct, this would completely remove Webpack/Webpacker from the codebase.

> Before starting, can we validate that we are able to replace all existing webpack functionality with vite?

I think so, are there any very specific plugins / webpack configs used in Mastodon?

I see you are generating packs for each locale, this can easily be replaced with [glob imports](https://vitejs.dev/guide/features.html#glob-import) in Vite.


> That for me needs to be put against the Shakapacker solution, for which I at least imagine, there might be less work in total

Moving to Shakapacker would definitely be easier, I will not lie on this. But I think there are real benefits to move away from Webpack.

> I do not know how (technical) decisions like this are taken
This is a good question 😉

**Comment by renchap** [comment]
I spent a few hours trying to migrate to ViteJS, results are here: https://github.com/renchap/mastodon/tree/vitejs

I put everything in 1 big commit, but this is just an exploratory PoC and, if I have to do it properly, I will split each different step in various commits (and probably start some cleanup PRs before the proper Vite migration PR).

Some things I noticed along the way:
- locale loading is very tied to Webpack and might need to be redone
- there are still `require()` and `module.exports =` in the codebase, this is not encouraged and will probably be tackled before any migration (even to webpack 5)
- JSX files have a `.js`. This is not a best-practice, but is quite easy to change (and does not require ViteJS)

I think switching to ViteJS is doable and has advantages: support for newer NodeJS versions, more supported tooling, far less configuration needed, very nice features (glob imports for example), much smaller JS output, and much faster build (a production build with Vite takes 15s on my laptop).

I can invest more time in this, but I would like a go from a project maintainer before doing so!

**Comment by olof-nord** [comment]
I think we have enough info to start work on a decision now. 

Now as for getting in touch with project maintainers, I did look into the people [listed](https://github.com/orgs/mastodon/people) in the GitHub Mastodon organization and found @Gargron - the same person is the one which has the highest commit count to this repo. 

@Gargron, how are (technical) decisions of this size made, and what is your opinion on a way forward in this case?

**Comment by olof-nord** [comment]
I had a look at historical PRs which could be related to this issue.

Webpack 2 was introduced in 2017, replacing sprockets/browserify.
This work was coordinated by @Gargron.
https://github.com/mastodon/mastodon/pull/2617

The upgrade to Webpack 3 was carried out later the same year.
That work was done by @nolanlawson.
https://github.com/mastodon/mastodon/pull/3912

The upgrade to Webpack 4 followed in 2018.
This work was carried out by @ykzts.
https://github.com/mastodon/mastodon/pull/6655

Webpack 5 was released in 2020-10-10.
https://webpack.js.org/blog/2020-10-10-webpack-5-release/

**Comment by Fluqzy** [comment]
As Nodejs v18 is now standard in debian 12 "bookworm" i would suggest to bring this request up again. I know the requirements are debian 11, but since there is a new release of debian, there will be more focus on debian 12 and more and more less on debian 11, which might bring up security issues. 
Might be a good thing to update to nodejs v18, since theres already a lts version :)

**Comment by renchap** [comment]
Our current frontend stack, based on webpack 4, does not support Node 18+.

We are aware this is an issue, and the plan is to replace it with Vite: #24981

**Comment by vmstan** [comment]
Mastodon 4.2 supports Node 18/20. Closing as fixed.

**Comment by olof-nord** [comment]
Thanks for doing this! I notice a lot of recent tooling and build improvements. For the curious, I see that webpack was kept, with some iterative improvements and one forked lib (?). I didn't look into the ruby part, as I don't have that much domain knowledge there.

https://github.com/mastodon/mastodon/pull/26830
